### PR TITLE
Free angle ch

### PR DESCRIPTION
**Pull request by PierreLaurentinCS**
# Description of the problem

- 1 : For some specific problem, the angle of contact should not be specified which leaves one more term to solve in the assembly on the boundary. 
- 2 : No tool was available to compute the barycenter of one fluid in a multiphase simulations in the Cahn-Hilliard solver

# Description of the solution

- 1: A new free_angle boundary condition was added to the model for this to be solved.
- 2:  A new tool allowing to compute the barycenter of an object was added for Cahn-Hilliard simulations. This was useful for the rising bubble case.

# How Has This Been Tested?

- The barycenter function was used to compare the rising bubbles using Cahn-Hilliard ans VOF. Its usage will be demonstrated in a subsequent PR

# Documentation

- Boundary conditions multiphysics (boundary_conditions_multiphysics.rst) ans post-processing (post_processing.rst) sections were updated.

# Comments

- Anything that you want to add up that didn't fit in any other category can be written here

### CODE REVIEW

**Review by AmishgaAlphonius — CHANGES_REQUESTED** [review]
Nice work! :)
I have a few comments and suggestions here and there.

File: source/solvers/cahn_hilliard.cc [hunk 1]
```diff
@@ -522,6 +532,135 @@ CahnHilliard<dim>::postprocess(bool first_iteration)
...
+                this->simulation_control->get_current_time()};
+
+              TableHandler table = make_table_scalars_tensors(
+                time,
+                independent_column_names,
+                position_and_velocity_vectors,
+                dependent_column_names,
+                this->simulation_parameters.simulation_control.log_precision);
+
+              std::cout << "+------------------------------------------+"
```
Comment (voferreira) [inline source/solvers/cahn_hilliard.cc:618]: You could use  announce_string from utilities.h here. It is merely cosmetic, but eventually someone will need to change it if you do not implement in this version.

File: source/solvers/cahn_hilliard.cc [hunk 2]
```diff
@@ -886,6 +1025,111 @@ CahnHilliard<dim>::solve_linear_system(const bool initial_step,
...
+
+  return std::pair<Tensor<1, dim>, Tensor<1, dim>>(barycenter_location,
+                                                   barycenter_velocity);
+}
+
+template std::pair<Tensor<1, 2>, Tensor<1, 2>>
+CahnHilliard<2>::calculate_barycenter<TrilinosWrappers::MPI::Vector>(
+  const TrilinosWrappers::MPI::Vector &solution,
+  const TrilinosWrappers::MPI::Vector &current_solution_fd);
+
```
Comment (voferreira) [inline source/solvers/cahn_hilliard.cc:1116]:
```suggestion
```

File: source/solvers/cahn_hilliard_assemblers.cc [hunk 3]
```diff
@@ -374,6 +374,113 @@ CahnHilliardAssemblerAngleOfContact<dim>::assemble_rhs(
...
+                        {
+                          const double phi_potential_i =
+                            scratch_data.phi_potential[q][i];
+
+                          for (unsigned int j = 0; j < scratch_data.n_dofs; ++j)
+                            {
+                              const Tensor<1, dim> grad_phi_face_phase_j =
+                                scratch_data.grad_phi_face_phase[f][q][j];
+
+                              local_matrix(i, j) +=
```
Comment (voferreira) [inline source/solvers/cahn_hilliard_assemblers.cc:419]:
```suggestion
                              local_matrix(i, j) -=
```

File: source/solvers/cahn_hilliard_assemblers.cc [hunk 4]
```diff
@@ -374,6 +374,113 @@ CahnHilliardAssemblerAngleOfContact<dim>::assemble_rhs(
...
+                                -epsilon * epsilon * phi_potential_i *
```
Comment (voferreira) [inline source/solvers/cahn_hilliard_assemblers.cc:420]:
```suggestion
                                epsilon * epsilon * phi_potential_i *
```

**Review by voferreira — CHANGES_REQUESTED** [review]
Nice! I have some suggestions here and there but mostly cosmetics. Good job, Pierre!

**Review by voferreira — CHANGES_REQUESTED** [review]
@PierreLaurentinCS For the reasons we talked about, I am requesting changes again to make sure the rebasing procedure does not break the master branch on merging.

File: include/core/parameters.h [hunk 5]
```diff
@@ -756,10 +756,11 @@ namespace Parameters
...
     Parameters::FluidIndicator postprocessed_fluid;
 
-    // Enable barycenter calculation for fluid 1 in VOF simulations
-    bool calculate_vof_barycenter;
+    // Enable barycenter calculation for fluid 1 in VOF and Cahn-Hilliard
+    // simulations
+    bool calculate_barycenter;
 
-    // Prefix for the barycenter output
+    // Prefix for the vof and Cahn-Hilliard barycenter output
```
Comment (AmishgaAlphonius) [inline include/core/parameters.h:763]:
```suggestion
    // Prefix for the VOF and Cahn-Hilliard barycenter output
```

File: source/core/parameters.cc [hunk 6]
```diff
@@ -1592,16 +1592,16 @@ namespace Parameters
...
+        "Enable calculation of the barycenter location and velocity of the fluid 1 in VOF and Cahn-Hilliard simulations.");
 
       prm.declare_entry(
-        "VOF barycenter name",
-        "vof_barycenter_information",
+        "barycenter name",
+        "barycenter_information",
         Patterns::FileName(),
-        "File name output for the barycenter information in VOF simulations");
+        "Name of barycenter information output file in VOF or Cahn-Hilliard= simulations");
```
Comment (AmishgaAlphonius) [inline source/core/parameters.cc:1604]:
```suggestion
        "Name of barycenter information output file in VOF or Cahn-Hilliard simulations");
```

File: source/core/parameters.cc [hunk 7]
```diff
@@ -1592,16 +1592,16 @@ namespace Parameters
                         "Enable smoothing postprocessed vectors and scalars.");
 
       prm.declare_entry(
-        "calculate VOF barycenter",
+        "calculate barycenter",
         "false",
         Patterns::Bool(),
-        "Enable calculation of the barycenter location and velocity of the fluid 1 in VOF simulations.");
+        "Enable calculation of the barycenter location and velocity of the fluid 1 in VOF and Cahn-Hilliard simulations.");
```
Comment (AmishgaAlphonius) [inline source/core/parameters.cc:1598]:
```suggestion
        "Enable calculation of the barycenter location and velocity of fluid 1 in VOF and Cahn-Hilliard simulations.");
```

**Review by AmishgaAlphonius — COMMENTED** [review]
Last few comments :)

File: source/solvers/cahn_hilliard.cc [hunk 8]
```diff
@@ -399,11 +405,11 @@ CahnHilliard<dim>::calculate_phase_statistics()
   if (simulation_parameters.post_processing.verbosity ==
       Parameters::Verbosity::verbose)
     {
-      this->pcout << "Phase statistics : " << std::endl;
-      this->pcout << "\t     Min : " << min_phase_value << std::endl;
-      this->pcout << "\t     Max : " << max_phase_value << std::endl;
-      this->pcout << "\t Average : " << phase_average << std::endl;
-      this->pcout << "\t Integral : " << integral << std::endl;
+      this->pcout << "Phase statistics: " << std::endl;
```
Comment (voferreira) [inline source/solvers/cahn_hilliard.cc:408]: You could use the announce string with "-" as separator here to look like the VANS results.

Comment (voferreira) [reply source/solvers/cahn_hilliard.cc:408]: The results could be left aligned as well.

File: source/solvers/cahn_hilliard.cc [hunk 9]
```diff
@@ -399,11 +405,11 @@ CahnHilliard<dim>::calculate_phase_statistics()
...
+      this->pcout << "\t     Min: " << min_phase_value << std::endl;
+      this->pcout << "\t     Max: " << max_phase_value << std::endl;
+      this->pcout << "\t Average: " << phase_average << std::endl;
+      this->pcout << "\t Integral: " << integral << std::endl;
```
Comment (voferreira) [inline source/solvers/cahn_hilliard.cc:412]: Additionally, add a line at the end for fashion.

**Review by voferreira — APPROVED** [review]
Nothing else to say :)

**Review by AmishgaAlphonius — APPROVED** [review]
Looks good to me also :)

**Pull request merged**